validator_client/src/duties_service.rs

```diff
@@ -674,84 +682,74 @@ async fn poll_beacon_attesters_for_epoch<T: SlotClock + 'static, E: EthSpec>(
-            |beacon_node| async move {
-                let _timer = metrics::start_timer_vec(
-                    &metrics::DUTIES_SERVICE_TIMES,
-                    &[metrics::ATTESTER_DUTIES_HTTP_POST],
-                );
-                beacon_node
-                    .post_validator_duties_attester(epoch, local_indices)
-                    .await
-            },
-        )
-        .await
-        .map_err(|e| Error::FailedToDownloadAttesters(e.to_string()))?;
-
-    drop(fetch_timer);
-    let _store_timer = metrics::start_timer_vec(
-        &metrics::DUTIES_SERVICE_TIMES,
-        &[metrics::UPDATE_ATTESTERS_STORE],
-    );
+    // Request duties for either `INITIAL_DUTIES_QUERY_SIZE` validators or the count of validators for which we
+    // don't already know their duties for that epoch, whichever subset is larger. We use the `dependent_root`
+    // in the response to determine whether validator duties need to be updated. This is to ensure that we don't
+    // request for extra data unless necessary in order to save on network bandwidth.
+    let uninitialized_validators =
+        get_uninitialized_validators(duties_service, &epoch, local_pubkeys);
+    let indices_to_request = if uninitialized_validators.len() >= INITIAL_DUTIES_QUERY_SIZE {
+        uninitialized_validators.as_slice()
+    } else {
+        &local_indices[0..min(INITIAL_DUTIES_QUERY_SIZE, local_indices.len())]
+    };
 
+    let response =
+        post_validator_duties_attester(duties_service, epoch, indices_to_request).await?;
     let dependent_root = response.dependent_root;
 
-    // Filter any duties that are not relevant or already known.
-    let new_duties = {
+    // Find any validators which have conflicting (epoch, dependent_root) values or missing duties for the epoch.
+    let validators_to_update: Vec<_> = {
         // Avoid holding the read-lock for any longer than required.
         let attesters = duties_service.attesters.read();
-        response
+        local_pubkeys
+            .iter()
+            .filter(|pubkey| {
+                attesters.get(pubkey).map_or(true, |duties| {
+                    duties
+                        .get(&epoch)
+                        .map_or(true, |(prior, _)| *prior != dependent_root)
+                })
+            })
+            .collect::<Vec<_>>()
+    };
+
+    if validators_to_update.is_empty() {
+        // No validators have conflicting (epoch, dependent_root) values or missing duties for the epoch.
+        drop(fetch_timer);
+        return Ok(());
+    }
+
+    // Filter out validators which have already been requested.
+    let attesters_in_response: Vec<_> = response.data.iter().map(|duty| duty.pubkey).collect();
+    let indices_to_request = validators_to_update
+        .iter()
+        .filter(|pubkey| !attesters_in_response.contains(pubkey))
+        .filter_map(|pubkey| duties_service.validator_store.validator_index(pubkey))
+        .collect::<Vec<_>>();
```

Comment:
```suggestion
    let initial_duties = &response.data;
    let indices_to_request = validators_to_update
        .iter()
        .filter(|&&&pubkey| !initial_duties.iter().any(|duty| duty.pubkey == pubkey))
        .filter_map(|pubkey| duties_service.validator_store.validator_index(pubkey))
        .collect::<Vec<_>>();
```

It's a bit of a micro-optimisation, but you could avoid a heap allocation with this change.